code/core/src/shared/status-store/index.ts

```diff
@@ -0,0 +1,227 @@
+    getAll() {
+      return universalStatusStore.getState();
+    },
+    set(statuses) {
+      universalStatusStore.setState((state) => {
+        // Create a new state object to merge with the current state
+        const newState = { ...state };
+
+        // Process each status and merge it into the appropriate storyId record
+        for (const status of statuses) {
+          const { storyId, typeId } = status;
+
+          newState[storyId] = { ...(newState[storyId] ?? {}), [typeId]: status };
+        }
+        return newState;
+      });
+    },
+    onAllStatusChange(
+      listener: (
+        statuses: StatusesByStoryIdAndTypeId,
+        prevStatuses: StatusesByStoryIdAndTypeId
+      ) => void
+    ): ReturnType<typeof universalStatusStore.onStateChange> {
+      return universalStatusStore.onStateChange((state, prevState) => {
+        listener(state, prevState);
+      });
+    },
+    onSelect(listener) {
+      return universalStatusStore.subscribe(StatusStoreEventType.SELECT, (event) => {
+        listener(event.payload);
+      });
+    },
+    selectStatuses: (statuses: Status[]) => {
+      universalStatusStore.send({ type: StatusStoreEventType.SELECT, payload: statuses });
+    },
+    unset(storyIds?: StoryId[]): void {
+      // If no storyIds are provided, remove all statuses
+      if (!storyIds) {
+        universalStatusStore.setState({});
+        return;
+      }
+
+      universalStatusStore.setState((state) => {
+        const newState = { ...state };
+        for (const storyId in newState) {
+          if (storyIds.includes(storyId)) {
+            delete newState[storyId];
+          }
+        }
+        return newState;
+      });
+    },
+    typeId: undefined,
+  };
+
+  const getStatusStoreByTypeId = (typeId: StatusTypeId): StatusStoreByTypeId => ({
+    getAll: fullStatusStore.getAll,
+    set(statuses): void {
+      universalStatusStore.setState((state) => {
+        // Create a new state object to merge with the current state
+        const newState = { ...state };
+
+        // Process each status and merge it into the appropriate storyId record
+        for (const status of statuses) {
+          const { storyId } = status;
+          if (status.typeId !== typeId) {
+            // Validate that all statuses have the correct typeId
+            switch (environment) {
+              case 'server':
+                throw new ServerStatusTypeIdMismatchError({
+                  status,
+                  typeId,
+                });
+              case 'manager':
+                throw new ManagerStatusTypeIdMismatchError({
+                  status,
+                  typeId,
+                });
+              case 'preview':
+              default:
+                throw new PreviewStatusTypeIdMismatchError({
+                  status,
+                  typeId,
+                });
+            }
+          }
+
+          newState[storyId] = { ...(newState[storyId] ?? {}), [typeId]: status };
```

Comment:
Its a bit of a nit, and probably won't matter, but you're changing the order of keys here. A cleaner solution is to shallow copy the object first, then set the key on the copy after.